### PR TITLE
MADDPG-torch sample/predict action

### PR DESCRIPTION
**Pull request by ShuaibinLi**
add `predict` func, rewrite `sample`
add run_eval_episode

### CODE REVIEW

File: benchmark/torch/maddpg/train.py [hunk 1]
```diff
@@ -137,43 +132,33 @@ def train_agent():
...
 
-    t_start = time.time()
-    logger.info('Starting...')
+    total_steps = 0
+    total_episodes = 0
+    test_flag = 0
     while total_episodes <= args.max_episodes:
         # run an episode
         ep_reward, ep_agent_rewards, steps = run_episode(env, agents)
         summary.add_scalar('train_reward/episode', ep_reward, total_episodes)
```
Comment (rical730) [inline benchmark/torch/maddpg/train.py]: `'train_reward/episode'` -> `'train/episode_reward_wrt_episode'`

Comment (rical730) [inline benchmark/torch/maddpg/train.py]: please add a line to record reward with respect to total_steps
`summary.add_scalar('train/episode_reward_wrt_step', ep_reward, total_steps)`

File: parl/algorithms/torch/maddpg.py [hunk 2]
```diff
@@ -107,9 +78,30 @@ def __init__(self,
...
+                noted that in the discrete case we take the argmax along the last axis as action
+        """
+        policy = self.model.policy(obs)
+
+        if self.continuous_actions:
+            action = policy[0]
+        else:
+            action = F.softmax(policy, dim=-1)
+        if self.continuous_actions:
+            action = torch.tanh(action)
```
Comment (rical730) [inline parl/algorithms/torch/maddpg.py]: Why didn't you just place this line after line 94 ?

File: parl/algorithms/torch/maddpg.py [hunk 3]
```diff
@@ -123,12 +115,17 @@ def predict(self, obs, use_target_model=False):
...
-            action = torch.tanh(action)
+            random_normal = torch.randn(size=policy[0].shape).to(self.device)
+            action = policy[0] + torch.exp(policy[1]) * random_normal
+        else:
+            uniform = torch.rand_like(policy)
+            soft_uniform = torch.log(-1.0 * torch.log(uniform)).to(self.device)
+            action = F.softmax(policy - soft_uniform, dim=-1)
 
+        if self.continuous_actions:
+            action = torch.tanh(action)
```
Comment (rical730) [inline parl/algorithms/torch/maddpg.py]: Same problem here, it may be more clear to place this line after line 121

File: benchmark/torch/maddpg/train.py [hunk 4]
```diff
@@ -137,43 +132,33 @@ def train_agent():
...
-                               total_steps)
-            summary.add_scalar('use_time/1000episode', use_time,
-                               total_episodes)
+
+        # evaluate agents
+        if (total_episodes + 1) // args.test_every_episodes >= test_flag:
+            while (total_episodes +
+                   1) // args.test_every_episodes >= test_flag:
+                test_flag += 1
+                eval_episode_rewards, eval_episode_steps = run_evaluate_episodes(
```
Comment (rical730) [inline benchmark/torch/maddpg/train.py]: I think there must be something wrong with the indent

File: parl/algorithms/torch/maddpg.py (hunk 2 above)
Comment (ShuaibinLi) [reply parl/algorithms/torch/maddpg.py]: modified

**Review by rical730 — DISMISSED** [review]
LGTM

**Review by rical730 — APPROVED** [review]
LGTM~
